gmso/tests/test_units.py

```diff
@@ -0,0 +1,223 @@
+import re
+
+import pytest
+import unyt as u
+
+from gmso.tests.base_test import BaseTest
+from gmso.utils.misc import unyt_to_hashable
+from gmso.utils.units import LAMMPS_UnitSystems
+
+
+class TestUnitHandling(BaseTest):
+    def test_unyt_to_hashable(self):
+        hash(unyt_to_hashable(None))
+        hash(unyt_to_hashable(1 * u.nm))
+        hash(unyt_to_hashable([4, 4] * u.nm))
+
+        assert hash(unyt_to_hashable(1 * u.nm)) == hash(
+            unyt_to_hashable(10 * u.angstrom)
+        )
+        assert hash(unyt_to_hashable(1 * u.kg)) == hash(
+            unyt_to_hashable(1000 * u.g)
+        )
+
+        assert hash(unyt_to_hashable(1 * u.nm)) != hash(
+            unyt_to_hashable(1.01 * u.nm)
+        )
+        assert hash(unyt_to_hashable(1 * u.nm)) != hash(
+            unyt_to_hashable(1.01 * u.second)
+        )
+        assert hash(unyt_to_hashable(1 * u.nm)) != hash(
+            unyt_to_hashable([1, 1] * u.nm)
+        )
+
+    """
+    Utilities to make
+    [a] register units needed for unit systems
+    [a] need to be able to generate unit systems
+    [a] take a unyt and check for energy
+    [a] take a unyt and check for electron volts
+    [?] take a unyt and look for thermal energy
+    [a] get all dimensions from a unit
+    [a] convert a unit using a base system
+    [a] return a rounded float for a unit
+    [] be able to write out a unit without the conversion
+    [] attach units to a float with a unit system and dimensions
+    [] should have associated somewhere to handle units outside of unit system
+    # units should be done as a top level conversion
+    # need to implement this module into the writers
+    # can probably subclass some functions out of LAMMPS_UnitSystems
+
+    Tests to make
+
+    """
+
+    def test_unit_conversion(self):
+        usys = LAMMPS_UnitSystems("real")
+        parameter = 0.001 * u.Unit("mm")
+        n_decimals = 5
+        outStr = usys.convert_parameter(parameter, n_decimals=n_decimals)
+        assert float(outStr) == 10000.00000
+        assert outStr[::-1].find(".") == n_decimals
+
+        parameter = 1 * u.Unit("nm")
+        n_decimals = 5
+        outStr = usys.convert_parameter(parameter, n_decimals=n_decimals)
+        assert float(outStr) == 10.00000
+        assert outStr[::-1].find(".") == n_decimals
+
+    def test_unit_rounding(self):
+        usys = LAMMPS_UnitSystems("real")
+        parameter = 0.001 * u.Unit("nm")
+        n_decimals = 5
+        outStr = usys.convert_parameter(parameter, n_decimals=n_decimals)
+        assert outStr[::-1].find(".") == n_decimals
+
+    def test_unitsystem_setup(self):
+        usys = LAMMPS_UnitSystems("real")
+        assert usys.system.name == "lammps_real"
+
+        usys = LAMMPS_UnitSystems("lj", registry=u.UnitRegistry())
+        assert usys.system.name == "lj"
+
+    def test_get_dimensions(self):
```

Comment:
Do we also want to have individual test for `_dimensions_to_energy()`, `_dimensions_to_charge`, and `_dimensions_from_thermal_to_energy` individually? May make it easier to trouble later if any one of them fails.